presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -219,47 +219,49 @@ protected RelationPlan visitJoin(Join node, Void context)
-        Expression postInnerJoinCriteria = new BooleanLiteral("TRUE");
+        List<Expression> nonComparisonExpressions = new ArrayList<>();
+        List<Expression> postInnerJoinConditions = new ArrayList<>();
         if (node.getType() != Join.Type.CROSS && node.getType() != Join.Type.IMPLICIT) {
             Expression criteria = analysis.getJoinCriteria(node);
 
             RelationType left = analysis.getOutputDescriptor(node.getLeft());
             RelationType right = analysis.getOutputDescriptor(node.getRight());
-            List<Expression> leftExpressions = new ArrayList<>();
-            List<Expression> rightExpressions = new ArrayList<>();
-            List<ComparisonExpression.Type> comparisonTypes = new ArrayList<>();
+
+            List<Expression> leftComparisonExpressions = new ArrayList<>();
+            List<Expression> rightComparisonExpressions = new ArrayList<>();
+            List<ComparisonExpression.Type> joinConditionComparisonTypes = new ArrayList<>();
+
             for (Expression conjunct : ExpressionUtils.extractConjuncts(criteria)) {
                 conjunct = ExpressionUtils.normalize(conjunct);
-                if (!(conjunct instanceof ComparisonExpression)) {
-                    throw new SemanticException(NOT_SUPPORTED, node, "Unsupported non-equi join form: %s", conjunct);
-                }
-
-                ComparisonExpression comparison = (ComparisonExpression) conjunct;
-                ComparisonExpression.Type comparisonType = comparison.getType();
-                if (comparison.getType() != EQUAL && node.getType() != INNER) {
+                if (!isEqualComparisonExpression(conjunct) && node.getType() != INNER) {
                     throw new SemanticException(NOT_SUPPORTED, node, "Non-equi joins only supported for inner join: %s", conjunct);
                 }
-                Set<QualifiedName> firstDependencies = DependencyExtractor.extractNames(comparison.getLeft(), analysis.getColumnReferences());
-                Set<QualifiedName> secondDependencies = DependencyExtractor.extractNames(comparison.getRight(), analysis.getColumnReferences());
-
-                Expression leftExpression;
-                Expression rightExpression;
-                if (firstDependencies.stream().allMatch(left.canResolvePredicate()) && secondDependencies.stream().allMatch(right.canResolvePredicate())) {
-                    leftExpression = comparison.getLeft();
-                    rightExpression = comparison.getRight();
-                }
-                else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && secondDependencies.stream().allMatch(left.canResolvePredicate())) {
-                    leftExpression = comparison.getRight();
-                    rightExpression = comparison.getLeft();
-                    comparisonType = flipComparison(comparisonType);
+
+                if (isComparisionExpression(conjunct)) {
```

Comment:
I think this code can be simplified a lot by only using this code path for `=` and treating all other `ComparisonExpression` as `postInnerJoinConditions`